### PR TITLE
Disable stdin processing for all but single instance CLI. Fixes #1300

### LINKED ISSUE #1300: JNI Layer throws Exceptions when close method is called in parallel ON DIFFERENT MODELS
## Problem

In the JNI layer, when multiple passes are enabled (`> 1`) and an attempt is made to close separate models in parallel, exceptions can be thrown.  This is true even though each model has it's own lock to guard all accesses to the native code paths.  Only a **global** lock around calls to the model `close` methods seems to avoid this issue.

I'm looking for help on identifying if any critical sections of the C code that can be guarded by a lock to avoid thread-safety issues.  I'm not asking for the C code to lock.  I just want help trying to figure out where to put the locks in the Java code that wraps the C code.

### Scope

This seems to be in PR #1291 but was not fixed with PR #1295

## Discussion

From empirical testing, it appears one of these lines seems to be the problem.  I am wondering if any of these use global state.  I am trying to figure out if we can lock only over a short critical section to avoid thread-safety issues.

1. ` adjust_used_index(*vwInstance);`
1. `vwInstance->do_reset_source = true;`
1. `VW::start_parser(*vwInstance);`
1. `LEARNER::generic_driver(*vwInstance);`
1. `VW::end_parser(*vwInstance);`

## Previous Conversation
In PR #1295 there was the following conversation:

@JohnLangford 

>There should be zero shared state between multiple created VW objects.  Is
that what it's doing?  (Creating multiple distinct VW objects?)

@deaktator 

> @JohnLangford. It looks like just one VW object. Each Java call does the following on the C side:
>
> `vw* vwInstance = VW::initialize(env->GetStringUTFChars(command, NULL);`

@JohnLangford 

>A single VW object can not be operated on in multiple threads because the code inside VW is not thread safe.  If you want to have a model which is shared by multiple threads, you set this up more explicitly by initializing a new VW object with an existing model.

@deaktator

> Hey @JohnLangford. We take care of multi-threaded access to VW by locking anywhere that requires access to the C code. The thread-safety issues I encountered before were on an incomplete version of the code that locked in the wrong place. When I run the tests in parallel, they seem to work just fine now. I ran them a bunch of times with forking in the tests and didn't see any issues.

## Tracking Down What's Happening

It appears @jon-morra-zefr pretty much copied the C# code for multiple passes, so this seems like it might apply to C# as well.  Both C# and JNI C++ code appear below as well as the calling code that blows up.  

I've seen a bunch of different errors that occur at the same spot.  invalidated cache, malformed LDF feature exceptions, etc.

## Example Code That triggers exceptions

```scala
// Doing this many times in parallel with no locks causes problems.

	val vwJNI = VWLearners.create[VWTypedLearner[_]](vwLearnString)
	// Learning in here using   vwJNI.learn

	// PROBLEM AREA:
	lock.lock()     // <== NEED GLOBAL LOCKING OR EXCEPTIONS THROWN
	vwJNI.close()
	lock.unlock()   // <== NEED GLOBAL LOCKING OR EXCEPTIONS THROWN
```

### C# Code: vowpal_wabbit/cs/cli/vowpalwabbit.cpp 
```csharp
void VowpalWabbit::RunMultiPass()
{ if (m_vw->numpasses > 1)
  { try
    { adjust_used_index(*m_vw);
      m_vw->do_reset_source = true;
      VW::start_parser(*m_vw);
      LEARNER::generic_driver(*m_vw);
      VW::end_parser(*m_vw);
    }
    CATCHRETHROW
  }
}
```

### JNI C++ Code: vowpal_wabbit/java/src/main/c++/vowpalWabbit_learner_VWLearners.cc

```c
JNIEXPORT void JNICALL Java_vowpalWabbit_learner_VWLearners_performRemainingPasses(JNIEnv *env, jclass obj, jlong vwPtr)
{ try
  { vw* vwInstance = (vw*)vwPtr;
    if (vwInstance->numpasses > 1)
      { adjust_used_index(*vwInstance);
        vwInstance->do_reset_source = true;
        VW::start_parser(*vwInstance);
        LEARNER::generic_driver(*vwInstance);
        VW::end_parser(*vwInstance);
      }
  }
  catch(...)
  { rethrow_cpp_exception_as_java_exception(env);
  }
}
```

*Any thoughts?*

### PR DESCRIPTION
**Pull request by kumpera**
Fixes #1300

### CODE REVIEW

**Comment by JohnLangford** [comment]
Do you know what the travis fail is?

**Comment by kumpera** [comment]
I'll take a look at it asap.

File: java/src/main/java/vowpalWabbit/learner/VWLearners.java [hunk 1]
```diff
@@ -36,6 +36,8 @@ private VWLearners() {}
      */
     @SuppressWarnings("unchecked")
     public static <T extends VWLearner> T create(final String command) {
+        if(command.indexOf("--no_stdin") == -1)
+            command += " --no_stdin";
```
Comment (jackgerrits) [inline java/src/main/java/vowpalWabbit/learner/VWLearners.java]: Travis build failure: `/vw/java/src/main/java/vowpalWabbit/learner/VWLearners.java:[40,13] final parameter command may not be assigned`

`command` is final so can't be changed, can just create a new string to append to to fix.

**Comment by JohnLangford** [comment]
Status?

**Comment by JohnLangford** [comment]
Status on this?

**Comment by kumpera** [comment]
I'm swamped with other higher priority tasks. I plan to have time for this before EOM.

**Comment by JohnLangford** [comment]
Sounds good.

**Comment by kumpera** [comment]
@jackgerrits @JohnLangford I fixed the Java issue. Sorry for the delay

**Comment by JohnLangford** [comment]
Great to have this one in, thanks :-)

**Pull request merged**